src/Storages/MergeTree/MutateTask.cpp

```diff
@@ -1511,11 +1726,20 @@ bool MutateTask::prepare()
 
     /// All columns from part are changed and may be some more that were missing before in part
     /// TODO We can materialize compact part without copying data
-    if (!isWidePart(ctx->source_part)
-        || (ctx->mutation_kind == MutationsInterpreter::MutationKind::MUTATE_OTHER && ctx->interpreter && ctx->interpreter->isAffectingAllColumns()))
+    if (need_mutate_all_columns)
     {
         task = std::make_unique<MutateAllPartColumnsTask>(ctx);
     }
+    else if (ctx->is_lightweight_mutation)
+    {
+        ctx->files_to_skip = ctx->source_part->getFileNamesWithoutChecksums();
+        /// Skip to create hardlink for deleted_rows_mask.bin
```

Comment:
I thought that `files_to_skip` means "files to hard link, not to mutate", but ok, maybe I'm wrong.

Reply:
files_to_skip means files not to hard link, like, checksums.txt, columns.txt ... 
Referenced from MutateSomePartColumnsTask's prepare():
...
        NameSet hardlinked_files;
        /// Create hardlinks for unchanged files
        for (auto it = ctx->source_part->data_part_storage->iterate(); it->isValid(); it->next())
        {
            if (ctx->files_to_skip.contains(it->name()))
                continue;
....